codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/AccountIDEndpointRouting.java

```diff
@@ -0,0 +1,70 @@
+package software.amazon.smithy.aws.go.codegen.customization;
+
+import software.amazon.smithy.aws.go.codegen.SdkGoTypes;
+import software.amazon.smithy.codegen.core.SymbolProvider;
+import software.amazon.smithy.go.codegen.*;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.rulesengine.traits.EndpointRuleSetTrait;
+import software.amazon.smithy.utils.MapUtils;
+
+import static software.amazon.smithy.go.codegen.GoWriter.goTemplate;
+
+public class AccountIDEndpointRouting implements GoIntegration {
+    @Override
+    public void renderPreEndpointResolutionHook(GoSettings settings, GoWriter writer, Model model) {
+        writer.write("""
+                if err := checkAccountID(getIdentity(ctx), m.options.AccountIDEndpointMode); err != nil {
+                    return out, metadata, $T("invalid accountID set: %w", err)
+                }
+                """,
+                GoStdlibTypes.Fmt.Errorf);
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        if (!settings.getService(model).hasTrait(EndpointRuleSetTrait.class)) {
+            return;
+        }
+        goDelegator.useShapeWriter(settings.getService(model), goTemplate("""
+        func checkAccountID(identity $auth:T, mode $accountIDEndpointMode:T) error {
+            switch mode {
+            case $aidModeUnset:T:
+            case $aidModePreferred:T:
+            case $aidModeDisabled:T:
+            case $aidModeRequired:T:
+                if ca, ok := identity.(*$credentialsAdapter:T); !ok {
+                    return $errorf:T("the accountID is configured to be required, but the " +
+                        "identity provider could not be converted to a valid credentials adapter " +
+                        "and provide an accountID, should switch to a valid credentials provider " +
+                        "or change to another account id endpoint mode")
```

Comment:
We don't have to specific about the type assertion failing in this path - we can just say an accountID wasn't found.

```suggestion
                    return $errorf:T("accountID is required but not set")
```

(use this value in the next branch as well)